### PR TITLE
PR#7315: refine some error locations

### PR DESCRIPTION
**Pull request by gasche**
Fixes [MPR#7315](http://caml.inria.fr/mantis/view.php?id=7315) and some related imprecise messages -- I suspect that more remain, and started building a testsuite.

### CODE REVIEW

**Comment by alainfrisch** [comment]
Ah, I was also working on it.

https://github.com/ocaml/ocaml/compare/trunk...alainfrisch:better_loc_unbound_constructor

I'd propose to change the interface of `Typetexp.find_type` (and then other `Typetexp.find_*` functions as well) to take a `longident Location.loc` instead of two separate arguments.  This encourage proper use on the call site, passing the exact location corresponding to the item being looked up.  This would catch, for instance, the weird location on:

``` ocaml
# type exn += X = Y;;
Characters 12-17:
  type exn += X = Y;;
              ^^^^^
Error: Unbound constructor Y
```

**Comment by gasche** [comment]
That sounds like a good idea. Feel free to integrate my tests, add some, and send a pull-request with your implementation.

**Comment by alainfrisch** [comment]
I'm not sure I'll have time to finalize that quickly, so it's perhaps better to merge this PR and keep the API change in mind for later.

Do you think the two failing tests are related to this PR?

```
    tests/typing-misc/polyvars.ml
    tests/typing-objects/pr6383.ml.reference
```

**Comment by gasche** [comment]
Yes, the tests were related to changed location information. I fixed the tests and integrated your own changes to extension redefinitions (plus lookup of the extensible type's name). I'll wait to leave some opportunity for review, then merge in trunk. Thanks for the advice!

File: testsuite/tests/typing-misc/polyvars.ml [hunk 1]
```diff
@@ -9,12 +9,12 @@ Error: This pattern matches values of type [? `A | `B ]
 |}];;
 let f x = ignore (match x with #ab -> 1); ignore (x : [`A]);;
 [%%expect{|
-Line _, characters 31-34:
+Line _, characters 32-34:
```
Comment (alainfrisch) [inline testsuite/tests/typing-misc/polyvars.ml]: One could argue that the pattern here is `#ab`,  not `ab`, so the old location might be better.

File: typing/typecore.ml [hunk 2]
```diff
@@ -1358,7 +1358,7 @@ let rec type_pat ~constrs ~labels ~no_existentials ~mode ~explode ~env
                   pat_extra = extra :: p.pat_extra}
         in k p)
   | Ppat_type lid ->
-      let (path, p,ty) = build_or_pat !env loc lid.txt in
+      let (path, p,ty) = build_or_pat !env lid.loc lid.txt in
```
Comment (alainfrisch) [inline typing/typecore.ml]: One should probably pass both `loc`, the location of the pattern (for errors related to the pattern); and `lid` (including `lid.loc`) to be able to report errors related to the lookup of this identifier.

Comment (gasche) [inline typing/typecore.ml]: Indeed, and this fixes the mistake you commented about above.

File: testsuite/tests/typing-objects/pr6383.ml.reference [hunk 3]
```diff
@@ -1,6 +1,6 @@
 
-# Characters 10-16:
+# Characters 11-16:
   let f (x: #M.foo) = 0;;
-            ^^^^^^
+             ^^^^^
```
Comment (gasche) [inline testsuite/tests/typing-objects/pr6383.ml.reference:4]: Note: it's not possible to refine the location here to `M` only without adding more location information inside `Longident.t`, which would result in a painful change to the interface used by everyone manipulating the parsetree -- not a good idea.

**Comment by ghost** [comment]
The original behavior:

<pre>
# type foo = <b>(unit,unit,unit,unit) bar</b>;;
Error: Unbound type constructor bar
# type foo = <b>unit bar</b>;;
Error: Unbound type constructor bar
</pre>

The new behavior:

<pre>
# type foo = (unit,unit,unit,unit) <b>bar</b>;;
Error: Unbound type constructor bar
# type foo = unit <b>bar</b>;;
Error: Unbound type constructor bar
</pre>

It is not a big thing, but nevertheless a strict improvement, in my opinion.

**Pull request merged**